### PR TITLE
Allow macros in char classes

### LINKED ISSUE #216: Macros in character classes are not supported
Using a macro in a character class produces the error message “Macros in character classes are not supported.” This contradicts [the documentation](http://jflex.de/manual.html#Grammar):

> ```
> CharClassContent    ::= CharClass | Character |
>                         Character'-'Character | 
>                         MacroUsage | PredefinedClass
> ```
> a character class [...] matches any character in that class. A Character is considered an element of a class if it is listed in the class or if its code lies within a listed character range Character’-’Character or Macro or predefined character class.

### PR DESCRIPTION
**Pull request by lsf37**
fixes #216 

* refactor parsing of character classes to produce an AST
* extend macro expansion to that AST

This PR refactors and redesigns how parsing of char classes works, and by that removes some duplication in the .cup file. It also moves the definition of char classes/regular expressions like `empty`, `anyChar`, `newLine` out of the .cup file into `RegExp` and `IntCharSet` respectively, to ensure consistency.

Compound character classes (= classes that are not a primitive predefined class or unicode property) are now first fully parsed, then in a separate pass after macro expansion, normalised into one `IntCharSet` that fully describes the class. Only after that step, we partition the input character set into these classes. Apart from being much cleaner, this should in theory lead to a slightly better (=coarser) partition of the character set, because we don't make a partition for each operand in a compound expression separately. In practice, the effect of this is probably minimal.

### CODE REVIEW

**Comment by lsf37** [comment]
This is a fairly big PR, but in total it removes code, which I'm happy about, and it didn't require any changes to the test suite (at least not after the recent commit that ensured a consistent class order).

**Comment by lsf37** [comment]
This is still somewhat WIP. It should work now, but I intend to add at least one test case for macro expansion, and to make Codacy a bit happier.

**Comment by lsf37** [comment]
Alright, rebased and cleaned up. @regisd, @sarowe the PR is now ready for review.

**Comment by lsf37** [comment]
Thanks for the additional cleanup!

How strongly do you feel about using `map`? I had it in one place originally, but removed it again, because the imperative version is easy to read and seeing how awkward `map` is in Java every time compared to functional languages made me somewhat sad. It's not really critical either way from my side, though.

**Comment by regisd** [comment]
> How strongly do you feel about using `map`? I had it in one place originally, but removed it again, because the imperative version is easy to read and seeing how awkward `map` is in Java every time compared to functional languages made me somewhat sad. 

If you talk about java8 streams, I don't feel strong at all.

File: jflex/src/main/java/jflex/core/IntCharSet.java [hunk 1]
```diff
@@ -150,7 +179,7 @@ public void add(Interval interval) {
   }
 
   /**
-   * add.
+   * Add a single character.
```
[the remark this answers is not in the document]
Comment (regisd) [reply jflex/src/main/java/jflex/core/IntCharSet.java]: > Thanks for the additional cleanup!
> 
Sorry, I forgot to send the first-pass review in which I was warning you about that. I had taken care of merging this into your branch, so it should have been painless for you if you did a `git pull`.

[the remark this answers is not in the document]
Comment (lsf37) [reply jflex/src/main/java/jflex/core/IntCharSet.java]: All good, no problems on this side.

File: jflex/src/main/cup/LexParse.cup [hunk 2]
```diff
@@ -94,184 +89,179 @@ action code {:
...
-        }
-        case sym.UPPERCLASS: {
-          // "Uppercase" is more than Uppercase_Letter, but older Unicode
-          // versions don't have this definition - check for "Uppercase",
-          // then fall back to Uppercase_Letter (Lu) if it does not exist.
-          intCharSet = unicodeProperties.getIntCharSet("Uppercase");
-          if (null == intCharSet) {
-            intCharSet = unicodeProperties.getIntCharSet("Lu");
+        case sym.LETTERCLASS:
+          {
```
[the remark this answers is not in the document]
Comment (lsf37) [reply jflex/src/main/cup/LexParse.cup]: Done, as far as possible. Some of the blocks declare new local variables which should stay local to that block, but the rest is gone.

**Comment by lsf37** [comment]
> > How strongly do you feel about using `map`? I had it in one place originally, but removed it again, because the imperative version is easy to read and seeing how awkward `map` is in Java every time compared to functional languages made me somewhat sad.
> 
> If you talk about java8 streams, I don't feel strong at all.

I think I'll leave it in after all, it's not so bad.

**Comment by lsf37** [comment]
Will rebase manually to clean up the commits a bit (this is a not unlikely point for future bisects), and merge after.

**Pull request merged**